packages/@uppy/components/src/hooks/dropzone.ts

```diff
@@ -0,0 +1,124 @@
+import type { UppyContext } from '../types.js'
+
+export type DropzoneOptions = {
+  noClick?: boolean
+  onDragOver?: (event: Event) => void
+  onDragEnter?: (event: Event) => void
+  onDragLeave?: (event: Event) => void
+  onDrop?: (files: File[]) => void
+  onFileInputChange?: (files: File[]) => void
+}
+
+export type DropzoneReturn<DragEventType, ChangeEventType> = {
+  getRootProps: () => {
+    onDragEnter: (event: DragEventType) => void
+    onDragOver: (event: DragEventType) => void
+    onDragLeave: (event: DragEventType) => void
+    onDrop: (event: DragEventType) => void
+    onClick: () => void
+  }
+  getInputProps: () => {
+    id: string
+    type: 'file'
+    multiple: boolean
+    onChange: (event: ChangeEventType) => void
+  }
+}
+
+const fileInputId = 'uppy-dropzone-file-input' as const
+
+export function createDropzone<
+  DragEventType extends DragEvent,
+  ChangeEventType extends Event,
+>(
+  ctx: UppyContext,
+  options: DropzoneOptions = {},
+): DropzoneReturn<DragEventType, ChangeEventType> {
+  const handleDrop = (event: DragEventType) => {
+    event.preventDefault()
+    event.stopPropagation()
+
+    const files = Array.from(event.dataTransfer?.files || [])
+    if (!files.length) return
+
+    if (options.onDrop) {
+      options.onDrop(files)
+    }
+
+    ctx.uppy?.addFiles(
```

Comment:
what's the use case of having ctx.uppy nullish but not crash? why do we want the code work but to silently not add any files? wouldn't it be better to crash so that the bug can be easier to identify?
```suggestion
    ctx.uppy!.addFiles(
```